dali/util/file.cc

```diff
@@ -16,15 +16,27 @@
 
 #include "dali/util/file.h"
 #include "dali/util/mmaped_file.h"
-#include "dali/util/std_file.h"
 #include "dali/util/odirect_file.h"
+#include "dali/util/std_file.h"
+#include "dali/util/uri.h"
+
+#if AWSSDK_ENABLED
+#include "dali/util/s3_client_manager.h"
+#include "dali/util/s3_file.h"
+#endif
 
 namespace dali {
 
-std::unique_ptr<FileStream> FileStream::Open(const std::string& uri, bool read_ahead,
-                                             bool use_mmap, bool use_odirect) {
-  std::string processed_uri;
+std::unique_ptr<FileStream> FileStream::Open(const std::string& uri, bool read_ahead, bool use_mmap,
+                                             bool use_odirect) {
+#if AWSSDK_ENABLED
+  if (uri.rfind("s3://", 0) == 0) {
+    return std::unique_ptr<FileStream>(
+        new S3FileStream(S3ClientManager::Instance().client(), uri));
+  }
+#endif
```

Comment:
I'm a bit worried about the flow. With the plain file handler, we close the file after reading (at least not keep too many files opened at the same time due to the limited numbers of FD the OS can maintain). In this case opening/closing stream can be expensive. Do you know how expensive/cheap are:
- creation of the `S3ClientManager`
- creation of the `S3FileStream`
Is it worth catching anything?

Reply:
we don't really open anything here. We query the API for the file size and store it. Then we keep a `pos_` and send a single Read request to the API every time there's a call to Read. There's no session open to the file.